### ISSUE TITLE
After uninstalling the module, the related tab in the BO menu is not found 

### ISSUE THREAD

**Issue by khouloudbelguith** [PrestaShop/PrestaShop#11264]
This issue has been edited because this is a general behaviour for all modules which have an entry in the BO menu (not only MBO).

See previous description below

**Describe the bug**
After uninstalling the "mbo" module, I tried to navigate to the Modules => Module Catalog page.
A red alert is displayed "Page not found" with an empty page.
We need to go to the Dashboard page & reload again the  Module Catalog page, to be OK.

**To Reproduce**

1. Go to Module Manager page
1. Uninstall the mbo module
1. Go to the Module Catalog page
1. See error

**Screenshots**
![image](https://user-images.githubusercontent.com/16067358/48001347-94f6f280-e108-11e8-9448-ed805379ce60.png)
https://drive.google.com/file/d/1t374pjGSzir34IPefLOshpgrNdCqd4UQ/view

**Additionnal information**
PrestaShop version: 1.7.5.0build4
PHP version: 7.1

**Comment by marionf** [comment]
@MrBaiame @eternoendless 
I don't know if this one should be fixed in the module or in the core ?

**Comment by jolelievre** [comment]
hello, this is not really a bug

this happens when you *just* uninstalled the module, because the link in the menu still directs you to the (uninstalled) module
the same problem happens for any uninstalled modules (link widget for instance), because we don't update the menu
And I am not sure we should for performance issues, but it could be a new feature

Anyway when you refresh your page the link is correctly reset to the core module manager

**Comment by marionf** [comment]
Will it be possible to reload the page when we click on uninstall ?

**Comment by jolelievre** [comment]
Yes it is possible, only it makes you lose your search if you made one
And it slows down the process if you were planning on uninstalling several modules (which doesn't happen every day I think)
But it kind of changes the feature, I don't know if we can decide this kind of changes on our own?

**Comment by marionf** [comment]
@colinegin wdyt ?

**Comment by colinegin** [comment]
Not sure I exactly understand, but in my opinion we should keep the catalog page even though MBO is uninstalled, is this possible @jolelievre  ?

**Comment by jolelievre** [comment]
Mmm, I am not sure to understand you either ^^

The native catalog page is always available, when you install MBO the link in the left side menu is replaced by the MBO catalog
So when you uninstall the module this link is no longer valid, but it is still in the menu
You just need to refresh the page for the menu to be updated with the native link

Hence our question, should we refresh the page each time a module is uninstalled in order to update the menu?

**Comment by colinegin** [comment]
Sorry I did not get what you meant :)

We discussed with @marionf , do you think it could have a bad impact on performance if we reload a page every time a module (or several ones simultaneously) is uninstalled ?

**Comment by jolelievre** [comment]
well it will take a few seconds for the page to be reloaded, is it acceptable?

**Comment by Quetzacoalt91** [comment]
You may uninstall several modules in a single call, but I would recommend to target 1.7.6.0 for coming back on a page refresh. The ajax requests have been implemented since PS 1.7.

**Comment by colinegin** [comment]
Seems a lot if done for every uninstalled module. 
Couldn't it be reloaded only when it detects MBO module ?

Otherwise can't we just simply add information on the page instead of the current message which is quite unclear ?

**Comment by LouiseBonnard** [comment]
I suggest: _Page cannot be found because a module has just been uninstalled. Please refresh the page, everything will be back in order!_ Let me know if I misunderstood the point, I can suggest other wordings. ;-)

**Comment by colinegin** [comment]
@NadiaCamard any news on the page new design ?

**Comment by eternoendless** [comment]
> Couldn't it be reloaded only when it detects MBO module ?

If I'm not mistaken, this will also happen with any module that adds elements to the menu, or elsewhere that's visible in the module manager (eg. icons on the top of the page, like gamification).

**Comment by colinegin** [comment]
Hello @eternoendless, 

Yes you are absolutely right, this happens on MBO and on every module which has a tab in the menu.
For this reason, I've updated the issue so that we can have a general decision on this topic.

There are 2 options : 
1- Create a dedicated custom error page with some information like : "Page cannot be found because a module has just been uninstalled. Please refresh the page, everything will be back in order!"
2- Autorefresh the page after uninstalling the module, with a message saying the module was uninstalled and a redirection to the module catalog page.

Which option suits you best @eternoendless @PrestaShop/prestashop-core-developers ?

Note : there will be a specific behaviour for MBO, we are currently discussing about what to do with it.

Thanks !

**Comment by jolelievre** [comment]
The thing is we have no way to know which modules have been installed/uninstalled
At least for now, so we would need to have some kind of history of the actions performed on modules

Besides showing a specific error page or redirecting to the module catalog seems more like a "patch", but from a UX point of view it is not ideal because for a short time we still have this incorrect link which shouldn't be present

I think the cleaner way to do this would be to refresh the page after a module action (this is not for uninstalling only, same problem exist for install|enable|disable|upgrade). But it would be too heavy and slow to perform this on every modules, some dont need it.

I propose a solution in-between, we could detect if some modification happened on tabs during the module action, if so then we refresh the page automatically. I'm not sure but I believe we can detect any ObjectModel modification with the hook system (to be confirmed) so this should be doable.

**Comment by Quetzacoalt91** [comment]
> 2- Autorefresh the page after uninstalling the module, with a message saying the module was uninstalled and a redirection to the module catalog page.

If we need to refresh after a ajax call, let's go back to a classic HTTP request instead. Displaying your success / error message will be made easier.

**Comment by eternoendless** [comment]
I agree that processing the request synchronously is the most appropriate approach here.

**Comment by jolelievre** [comment]
I'm not sure I understand your discussion about ajax and synchronous request?
You would like to change the behaviour so that modules action are performed via a direct request? no more ajax?
How do you deal with bulk actions then?

**Comment by eternoendless** [comment]
Yes, that's right.

Regarding bulk actions, I'm not sure that they are really, actually necessary in this context. But if they choose to uninstall 20 modules at once, they better have set their time_limit to something greater than 30...

**Comment by jolelievre** [comment]
I don't think we will be able to uninstall 20 modules in one request either, which is why ajax calls seem appropriate in this condition. But you're right with 20 parallel requests some might fail with timeout.
We could update the js to perform only 5 requests a the same time

And regarding the refresh problem, I think the "module API" could add a field in its json response indicating if the module needs a page refresh it is done after all bulk actions are over.

**Comment by eternoendless** [comment]
I don't think people will be uninstalling 20 modules every day.

Either way, if we choose to keep bulk uninstall, and the bulk action handles every uninstall sequentially in a separate request, we can keep it and then refresh the page at the end of the process. If, on the other hand, all actions are processed in a single request, then I suggest blocking it from running it unless `max_execution_time` is 0.

**Comment by jolelievre** [comment]
The current behavior is:
- every actions is performed by its own request
- all actions are launched in parallel with a parameter to prevent clearing the cache (with 20 requests some may be pending)
- except ONE of the requests which waits for the others to be over, then we perform this last request with clear cache enabled

Since all actions are performed separately I don't think we should have timeout problems (in mostest cases)
Anyway we could easily detect if an action requires to refresh the page an return it in the json response, if one of the module needs a refresh then we do so.

**Comment by colinegin** [comment]
I'm wondering : will it happen the same if the module is disabled only?

**Comment by jolelievre** [comment]
I'm not sure it depends how the module managed it, but imho a disabled module should not be visible any more.
But I have trouble understanding the difference between disable and uninstall. Except that disable doesn't delete its data it should do the same thing, so the tab will disappear as well.

**Comment by colinegin** [comment]
Yeah I agree with you, in my opinion the tab should also disappear.
I just want to be sure we take into account this specific case.

**Comment by jolelievre** [comment]
The thing is we can't be sure, it depends on how the module manages its `disabled` function I'm pretty sure most of them don't, or if they do they might forget to remove the tabs.

Which is why I proposed to check for tab modifications on any module actions (I think it's doable with the hooks `actionObjectTabUpdateAfter` and `actionObjectTabAddAfter`). This way wether the module manages it right or not, we know for sure when the tab have been modified and thus we know if we need to refresh.

**Comment by hibatallahAouadni** [comment]
Hello everyone,

I didn't manage to reproduce the issue with PS1.7.7.0, see the attached screen record below:

https://drive.google.com/file/d/1RZl1Oa0cGXDR-UGaaRQCt6gLBNOrGZ3R/view

As you can see, when I uninstalled the mbo module and tried to access to Module Catalog, the system redirects me to **Dashboard**. Then when I went to **Modules**>**Module Catalog**, I didn't encounter any issue.

Please check and feedback.
Thanks!

**Comment by khouloudbelguith** [comment]
Hi @hibatallahAouadni,

Yes, the error is not reproduced anymore with PS1770 and ps_mbo v2.0.1.
https://drive.google.com/file/d/1ZfgZaE2IQpnr95l2frx4i9XLzZi3e_mc/view?usp=sharing
But, I have a strange behavior:

1. Go to Module Manager page
2. Uninstall the ps_mbo module
3. Go to the Module Catalog page
4. Redirection to Dashboard page.

Thank you!

**Comment by hibatallahAouadni** [comment]
Well then, I believe we need the Product Team opinion.
Ping @PrestaShop/product-team Do you think that this issue still requires work or this current behavior is acceptable?

**Comment by LouiseBonnard** [comment]
Hi @khouloudbelguith and @hibatallahAouadni! If the bug is fixed, I suggest to close the ticket and open a new one for this new issue. I've looked into the specifications and found nothing about it... When uninstalling the MBO module or any other module, the user should be redirected to the Module Catalog page if he/she clicks on it. Assign me the issue once created, I'll add the specification line and label it as _to do_, thanks!

**Comment by hibatallahAouadni** [comment]
Thanks @khouloudbelguith for your confirmation :stuck_out_tongue: 
Thanks @LouiseBonnard for your quick specification :wink: 
I will colse this one and I already open a new one #22740
